### PR TITLE
Rh/rsc

### PR DESCRIPTION
**Pull request by laholstege**
1) we have a AI and UI state. the AI is created at the bottom of `actions.tsx` and provided on the client like context on `app/page.tsx` 
2) we don't need to pass around many props, we do `useUIState` on the client to add to the state. we use a server action `submitUserMessage` instead of an `onSubmit`. 
3) submitUserMessage is where the magic happens. we start streaming a UI for the entire array of ReactNodes for the messages. so first we add the user's message, then the empty bot message 
4) we start the async process to generate the initial system message and stream updates 
5) we actually start running the LLM and stream that text back, then close out all the streams 
6) on `main.tsx` the UI itself is rendered as this because it is just an array of nodes
```
{messagesUI .map((m) => {return <div key={m.id}>{m.display}</div>;})}
```
---
things i'm not sure about: 
1) why session is fetched multiple times. my thought is that each page should check auth on the server before even calling the db, so i don't check auth() again on `getChat` or `getChatHistory`. i feel like this shouldn't matter since we're just doing the historyService call from the server..?
2) why it takes a second to load the 👍 👎  on feedback when you refresh the page. the UCD logo also flickers, and homepage policybot svg. also why the first load is so massive?? i get layout.js at 2.1 MB and page.js at 2.0 MB. if i comment out the font awesome icons/imports that goes down to 642 kB and 1.0 MB
3) we could show the message on the sidebar before it is completed (and an animation to make it look like it's typing) if we save the system (and possibly user) message to the db immediately, so when we do `router.refresh()` and the sidebar calls `getChats()` it will pull that one
4) better way to reset chat than going to `/` which redirects to `/chat/new` when clicking "ask a new question"
5) best pattern for error handling? i've done it on `page.tsx` so it handles not found or not authorized. but should those be errors thrown by the db? also i stopped using `Suspense` because the only real loading state we have is handled by the streamableUI. 
6) it's not super clear when it is unsafe to use data without sanitizing it. like, i'm calling `saveChat()` from actions.tsx (which is a server component)–could i just pass in the entire object? would it be safe to save the `llmModel` from that object?  

and i need @cydoval to look at the animations and also make the side history scrollable

## Summary by CodeRabbit

- **New Features**
  - Introduced animation effects for chat history.
  - Added a new `NotFound` component to display a 404 error message.
  - Added a `NotAuthorized` component for access restriction scenarios.
  - Implemented new chat box form for user interactions.
  - Enhanced chat application with `newChatSession` logic for better session management.

- **Enhancements**
  - Improved layout and presentation of the About page.
  - Updated chat components for better performance and user experience.
  - Reorganized chat service code for OpenAI integration.

- **Bug Fixes**
  - Fixed attribute naming conventions in SVG components.

- **Documentation**
  - Added necessary import statements across various components for clarity.

- **Style**
  - Added and updated styling for the chat history section.

- **Refactor**
  - Streamlined code and logic across chat components and services.

- **Dependencies**
  - Added `"framer-motion"` to enhance UI with motion effects.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -28,7 +28,8 @@
     "react-dom": "^18",
     "react-markdown": "^9.0.1",
     "reactstrap": "^9.2.2",
-    "remark-gfm": "^4.0.0"
+    "remark-gfm": "^4.0.0",
+    "zod": "^3.22.4"
```
Comment (srkirkland) [inline package.json]: i thought we liked `yup`, is `zod` the one now?

File: src/components/chat/chatMessage.tsx [hunk 2]
```diff
@@ -1,25 +1,68 @@
...
+    <div className='row mb-3'>
+      <div className='col-3 col-md-1 mb-2'>
+        <WonkPortrait isLoading={isLoading} />
+      </div>
+      <div className='col-10 col-md-11'>
+        <p className='chat-name'>
+          <strong>Policy Wonk</strong>
+        </p>
+        <div>
+          {wonkText}
```
Comment (srkirkland) [inline src/components/chat/chatMessage.tsx]: I like this but `wonkText` should disappear once it's done loading

File: src/lib/actions.tsx [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+async function submitUserMessage(userInput: string) {
+  'use server';
+
+  const userMsgId = nanoid();
+  const wonkMsgId = nanoid();
+
+  // before we actually do anything, stream loading UI (for the chat window)
+  // would be better to add this on the client side so it is immediate
+  // but putting it here for now
+  const chatWindowUI = createStreamableUI(
```
Comment (srkirkland) [inline src/lib/actions.tsx]: this is from ai lib, is streamableUI specific to the AI library or is this a common RSC pattern we could do with other libs?

File: src/lib/actions.tsx [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  // provided by <AI> in the page.tsx
+  const aiState = getMutableAIState<typeof AI>();
+
+  // then start our async process
+  // this is an immediately invoked function expression (IIFE) so that the above code is not blocked
+  (async () => {
+    wonkThoughts.update('Getting embeddings...');
+    const embeddings = await getEmbeddings(userInput);
+
+    wonkThoughts.update('Searching for relevant documents...');
```
Comment (srkirkland) [inline src/lib/actions.tsx:63]: now this is cool

**Review by srkirkland — COMMENTED** [review]
neat.  I like it, but the complexity is a tough bar to jump over for everyday use.  It's a better UX w/ the streaming though so I say go for it once reconciled w/ `main`

File: src/lib/actions.tsx (hunk 3 above)
Comment (laholstege) [reply src/lib/actions.tsx]: @srkirkland createStreamableUI is specific to ai/rsc, but it's just using suspense and promises underneath. what you get from SSR is the ability to stream components in chunks–the chunks being determined by being async components or wherever you specify a suspense boundary. in next these are automatically set at the page level, which is why you can do `loading.js` and `error.js`  

https://github.com/vercel/ai/blob/e3206ff19b850d40dd9cbbe53d9b7a93418bf622/packages/core/rsc/streamable.tsx#L24
https://github.com/vercel/ai/blob/e3206ff19b850d40dd9cbbe53d9b7a93418bf622/packages/core/rsc/utils.tsx#L52

suspense works by throwing promises and using the status to determine when to switch states for rendering. like: 
```js
  if (fetchRequest.status === 'pending') {
    throw fetchRequest;
  }
  ```
🤯 
so while it's pending, it uses whatever is supplied as the fallback. so looks like they're just not resolving that outer promise until you call like `ui.done()`. and within that, to update the actual react node before you call `.done()`, they are also leveraging suspense boundaries and promisifying components. 

so to update a node: 
```js
export function createStreamableUI(initialValue?: React.ReactNode) {
 // create outer suspense boundary and promise
  let { row, resolve, reject } = createSuspensedChunk(initialValue);
  ...
  return {
  ...
    append(value: React.ReactNode) {
      assertStream('.append()');

      const resolvable = createResolvablePromise();
      currentValue = value;

      // resolve current inner promise, but make a new one inside a suspense boundary 
      // which will display our currentValue until it is resolved (via 
      resolve({ value, done: false, append: true, next: resolvable.promise }); 
      resolve = resolvable.resolve;
      reject = resolvable.reject;
      ...

 ```
 
 and that resolve in append moves the suspense boundary around a new inner promise
 ```js
const Row = [
  (async ({ currrentNode, nextPromise}) => {
    const chunk = await nextPromise;
    if (chunk.done) {
      return chunk.value;
    }     
    if (chunk.append) {
      return (
        <>
          // what we had before appending, if !chunk.append this is just discarded 
          {currentNode} 
          // make a new suspense boundary but make the fallback the ReactNode 
          // (this is the wildest part to me) 
          // so the actual ReactNode we are rendering is actually rendered as the fallback
          //  BEFORE the promise is resolved
          //  and when we are resolving the promise that represents the currentNode
          // we create the next suspended node, suspended on the next promise
          <Suspense fallback={chunk.value}> 
            <Row currentNode={chunk.value} nextPromise={chunk.next} />
          </Suspense>
        </>
      );
      ...
    } 
```
 
 so tl;dr yes we can do that with just RSC (though i imagine there are/will be libraries to abstract this)

**Review by srkirkland — APPROVED** [review]

File: src/app/chat/[chatid]/page.tsx [hunk 5]
```diff
@@ -1,38 +1,66 @@
...
+import { llmModel } from '@/services/chatService';
 import { getChat } from '@/services/historyService';
 
 type HomePageProps = {
   params: {
     chatid: string;
   };
 };
-
+// TODO: loading animation when chatId changes
```
[the remark this answers is not in the document]
Comment (srkirkland) [reply src/app/chat/[chatid]/page.tsx:21]: open a github issue and assign it to laholstege.  Include some thoughts on how you would implement it

**Review by srkirkland — APPROVED** [review]

**Pull request merged**